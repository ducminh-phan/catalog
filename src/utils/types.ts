export type PaginationParams = {
  offset: number;
  limit: number;
};

export type Category = {
  id: number;
  name: string;
  description: string;
};
