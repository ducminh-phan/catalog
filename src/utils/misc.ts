export type JsonPrimitive = string | number | boolean | null;
export type JsonMap = {
  [member: string]: JsonPrimitive | JsonArray | JsonMap;
};
export type JsonArray = Array<JsonPrimitive | JsonArray | JsonMap>;
export type Json = JsonPrimitive | JsonMap | JsonArray;

export const toCamel = (str: string): string => {
  return str.replace(/([-_][a-z])/gi, ($1) => {
    return $1.toUpperCase().replace("-", "").replace("_", "");
  });
};

export const isObject = (obj: Json): obj is JsonMap => {
  return obj === Object(obj) && !Array.isArray(obj);
};

export const snakeToCamel = (obj: Json): Json => {
  if (isObject(obj)) {
    const n: { [key: string]: Json } = {};

    Object.keys(obj).forEach((k) => {
      n[toCamel(k)] = snakeToCamel(obj[k]);
    });

    return n;
  }

  if (Array.isArray(obj)) {
    return obj.map((i) => {
      return snakeToCamel(i);
    });
  }

  return obj;
};
