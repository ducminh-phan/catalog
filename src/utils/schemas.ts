import { schema } from "normalizr";

export const category = new schema.Entity("categories");

export const categories = [category];

export const item = new schema.Entity("items");

export const items = [item];
