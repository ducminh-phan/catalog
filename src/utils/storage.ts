class Storage {
  private readonly prefix: string;

  private static readonly tokenKey: "token";

  constructor(localStoragePrefix: string) {
    this.prefix = localStoragePrefix;
  }

  getItem = (key: string): string | null =>
    localStorage.getItem(`${this.prefix}:${key}`);

  setItem = (key: string, value: string): void =>
    localStorage.setItem(`${this.prefix}:${key}`, value);

  removeItem = (key: string): void =>
    localStorage.removeItem(`${this.prefix}:${key}`);

  getToken = (): string | null => this.getItem(Storage.tokenKey);

  setToken = (token: string): void => this.setItem(Storage.tokenKey, token);

  removeToken = (): void => this.removeItem(Storage.tokenKey);
}

export default new Storage(process.env.REACT_APP_LOCAL_STORAGE_PREFIX);
