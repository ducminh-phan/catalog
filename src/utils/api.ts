import { Json, snakeToCamel } from "./misc";
import storage from "./storage";

type Method = "get" | "post" | "put";

class Request {
  static request = async <T extends Json>(
    endpoint: string,
    method: Method,
    body?: object,
  ): Promise<T> => {
    const token = storage.getToken();
    const headers: HeadersInit = {
      Accept: "application/json",
    };

    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }

    const config: RequestInit = {
      method,
      headers,
    };

    if (body) {
      config.body = JSON.stringify(body);
      headers["Content-Type"] = "application/json";
    }

    const response = await window.fetch(
      `${process.env.REACT_APP_API_URL}/${endpoint}`,
      config,
    );

    const data = snakeToCamel(await response.json());

    if (response.ok) {
      return data as T;
    }

    return Promise.reject(data);
  };

  static get = <T extends Json>(endpoint: string): Promise<T> =>
    Request.request(endpoint, "get");

  static post = <T extends Json>(endpoint: string, body: object): Promise<T> =>
    Request.request<T>(endpoint, "post", body);

  static put = <T extends Json>(endpoint: string, body: object): Promise<T> =>
    Request.request<T>(endpoint, "put", body);
}

export default Request;
