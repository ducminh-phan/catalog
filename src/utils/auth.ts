import Request from "./api";
import storage from "./storage";

export interface LoginRequest {
  username: string;
  password: string;
}

interface LoginResponse {
  access_token: string;
}

export interface RegisterRequest {
  username: string;
  password: string;
  email: string;
  name: string;
}

interface RegisterResponse {
  access_token: string;
}

const handleAuthResponse = ({
  access_token: token,
}: LoginResponse | RegisterResponse): void => storage.setToken(token);

export const login = (body: LoginRequest): Promise<void> => {
  return Request.post<LoginResponse>("login", body).then(handleAuthResponse);
};

export const register = (body: RegisterRequest): Promise<void> => {
  return Request.post<RegisterResponse>("registrations", body).then(
    handleAuthResponse,
  );
};

export const logout = (): Promise<void> => {
  storage.removeToken();
  return Promise.resolve();
};

interface User {
  id: number;
  name: string;
}

export interface AppData {
  user: User;
}

interface UserResponse {
  email: string;
  username: string;
  created: string;
  updated: string;
  name: string;
  id: number;
}

export const getAppData = async (): Promise<AppData | null> => {
  const token = storage.getToken();

  if (token === null) {
    return null;
  }

  return Request.get<UserResponse>("me")
    .then(({ id, name }) => ({ user: { id, name } }))
    .catch((error) => {
      logout();
      return Promise.reject(error);
    });
};
