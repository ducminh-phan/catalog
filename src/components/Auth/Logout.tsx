import { Button } from "@material-ui/core";
import React, { ReactElement } from "react";

import { useAuth } from "contexts/auth";

const Logout = (): ReactElement => {
  const { logout } = useAuth();

  return (
    <Button color="inherit" onClick={logout}>
      Logout
    </Button>
  );
};

export default Logout;
