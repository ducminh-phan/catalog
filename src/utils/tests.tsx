import { StylesProvider } from "@material-ui/styles";
import { render as rtlRender, RenderResult } from "@testing-library/react";
import { Rule, StyleSheet } from "jss";
import React, { ComponentType, ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";

import Notification from "components/Notification";
import { AuthProvider } from "contexts/auth";
import { NotificationProvider } from "contexts/notification";
import store from "store";

import Request from "./api";
import { UserResponse } from "./auth";
import storage from "./storage";

// Custom logic to make Material UI class names consistent in snapshots
const generateClassName = (
  rule: Rule,
  styleSheet?: StyleSheet<string>,
): string => `${styleSheet?.options.classNamePrefix}-${rule.key}`;

const render = (ui: ReactElement, renderOptions = {}): RenderResult => {
  const Wrapper = ({ children }: { children: ReactNode }): ReactElement => {
    return (
      <StylesProvider generateClassName={generateClassName}>
        <Provider store={store}>
          <NotificationProvider>
            <AuthProvider>{children}</AuthProvider>
            <Notification />
          </NotificationProvider>
        </Provider>
      </StylesProvider>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper as ComponentType, ...renderOptions });
};

export const setupUser = (): void => {
  storage.setToken("accessToken");
  expect(storage.getToken()).toBe("accessToken");
  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<UserResponse> =>
      Promise.resolve({
        email: "minhπ",
        username: "minhπ",
        created: "",
        updated: "",
        name: "minhπ",
        id: 0,
      }),
  );
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
