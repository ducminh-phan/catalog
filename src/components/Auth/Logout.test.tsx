import React from "react";

import storage from "utils/storage";
import { fireEvent, render, setupUser, wait } from "utils/tests";

import Logout from "./Logout";

it("can logout", async () => {
  setupUser();

  const { baseElement, getByText } = render(<Logout />);

  await wait(() => getByText("Logout"));
  expect(baseElement).toMatchSnapshot();

  // Click the Login button
  await wait(() => fireEvent.click(getByText("Logout")));

  expect(storage.getToken()).toBeNull();
});
