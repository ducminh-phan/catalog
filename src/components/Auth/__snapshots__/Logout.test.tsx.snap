// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can logout 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Logout
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;
